### ISSUE TITLE
pyright ignores from __future__ import annotations when it comes to method definitions

### ISSUE THREAD

**Issue by fritsi** [microsoft/pyright#11236]
**Description**

This code works in Python 3.12, it even passed `mypy` validation, but fails with `pyright`:

```python
from __future__ import annotations

from typing import Annotated

from pydantic import BeforeValidator


type MyStr = Annotated[str, BeforeValidator(_my_before_validator)]


def _my_before_validator(value: str) -> str:
    return value
```

The error message is:

```
error: "_my_before_validator" is not defined (reportUndefinedVariable)
```

**Workaround:**

```python
type MyStr = Annotated[str, BeforeValidator(lambda v: _my_before_validator(v))]  # pylint: disable=unnecessary-lambda
```
